exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1321 @@
+    try {
+      sorter.setup(context, sv2, convertedBatch);
+    } catch (SchemaChangeException e) {
+      convertedBatch.clear();
+      throw UserException.unsupportedError(e)
+            .message("Unexpected schema change.")
+            .build(logger);
+    }
+    try {
+      sorter.sort(sv2);
+    } catch (SchemaChangeException e) {
+      throw UserException.unsupportedError(e)
+                .message("Unexpected schema change.")
+                .build(logger);
+    }
+    RecordBatchData rbd = new RecordBatchData(convertedBatch, allocator);
+    try {
+      rbd.setSv2(sv2);
+      bufferedBatches.add(new BatchGroup.InputBatch(rbd.getContainer(), rbd.getSv2(), oContext, batchSize));
+      if (peakNumBatches < bufferedBatches.size()) {
+        peakNumBatches = bufferedBatches.size();
+        stats.setLongStat(Metric.PEAK_BATCHES_IN_MEMORY, peakNumBatches);
+      }
+
+    } catch (Throwable t) {
+      rbd.clear();
+      throw t;
+    }
+  }
+
+  /**
+   * Update the data-driven memory use numbers including:
+   * <ul>
+   * <li>The average size of incoming records.</li>
+   * <li>The estimated spill and output batch size.</li>
+   * <li>The estimated number of average-size records per
+   * spill and output batch.</li>
+   * <li>The amount of memory set aside to hold the incoming
+   * batches before spilling starts.</li>
+   * </ul>
+   *
+   * @param actualBatchSize the overall size of the current batch received from
+   * upstream
+   * @param actualRecordCount the number of actual (not filtered) records in
+   * that upstream batch
+   */
+
+  private void updateMemoryEstimates(long actualBatchSize, int actualRecordCount) {
+
+    // The record count should never be zero, but better safe than sorry...
+
+    if (actualRecordCount == 0) {
+      return; }
+
+    // We know the batch size and number of records. Use that to estimate
+    // the average record size. Since a typical batch has many records,
+    // the average size is a fairly good estimator. Note that the batch
+    // size includes not just the actual vector data, but any unused space
+    // resulting from power-of-two allocation. This means that we don't
+    // have to do size adjustments for input batches as we will do below
+    // when estimating the size of other objects.
+
+    int batchRecordSize = (int) (actualBatchSize / actualRecordCount);
+
+    // Record sizes may vary across batches. To be conservative, use
+    // the largest size observed from incoming batches.
+
+    int origEstimate = estimatedRecordSize;
+    estimatedRecordSize = Math.max(estimatedRecordSize, batchRecordSize);
+
+    // Go no further if nothing changed.
+
+    if (estimatedRecordSize == origEstimate) {
+      return; }
+
+    // Maintain an estimate of the incoming batch size: the largest
+    // batch yet seen. Used to reserve memory for the next incoming
+    // batch.
+
+    estimatedInputBatchSize = Math.max(estimatedInputBatchSize, actualBatchSize);
```

Comment:
Need to move this line above the "return" on line 892 above -- Just in case a huge batch comes, but with a huge record count such that the average record does not exceed the previous estimate -- then we exit on line 892 without capturing that huge batch size.
  Actually the same applies to some more code below that makes use of the est. input batch size.